From 5183d5dffe1bc6f56eaa0a1d26ba59a10aa3456b Mon Sep 17 00:00:00 2001
From: Denis Vlasenko <vda.linux@googlemail.com>
Date: Tue, 26 Jun 2007 13:05:08 +0000
Subject: [PATCH] Restore pgrp on exit (fix "backgrounded MC" bug)

When I start dash under Midnight Commander and then type 'exit', dash
exits all right, but then MC is sent to background. It happens because
dash does not restore current process group on exit.

Attached patch fixes this. It also fixes another bug: setjobctl(0)
must ignore tcsetpgrp errors, because there are cases when tty is
destroyed under dash.

Patch is run-tested.
--
vda
---
 src/jobs.c |    5 ++++-
 src/trap.c |    3 ++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/jobs.c b/src/jobs.c
index 7285d0d..021640c 100644
--- a/src/jobs.c
+++ b/src/jobs.c
@@ -219,7 +219,10 @@ out:
 		/* turning job control off */
 		fd = ttyfd;
 		pgrp = initialpgrp;
-		xtcsetpgrp(fd, pgrp);
+		/* was xtcsetpgrp, but this is used on exit and
+		 * can loop forever if tty is already destroyed
+		 * (e.g. broken ssh link, closed xterm etc) */
+		tcsetpgrp(fd, pgrp);
 		setpgid(0, pgrp);
 		setsignal(SIGTSTP);
 		setsignal(SIGTTOU);
diff --git a/src/trap.c b/src/trap.c
index eae6186..00e3104 100644
--- a/src/trap.c
+++ b/src/trap.c
@@ -357,7 +357,7 @@ exitshell(void)
 	TRACE(("pid %d, exitshell(%d)\n", getpid(), status));
 	if (setjmp(loc.loc)) {
 		if (exception == EXEXIT)
-			_exit(exitstatus);
+			status = exitstatus;
 		goto out;
 	}
 	handler = &loc;
@@ -367,6 +367,7 @@ exitshell(void)
 	}
 	flushall();
 out:
+	setjobctl(0);
 	_exit(status);
 	/* NOTREACHED */
 }
-- 
1.5.2.2

